src/translator/processed_request_sentence.cpp

```diff
@@ -0,0 +1,102 @@
+#include "processed_request_sentence.h"
+
+#include <cstring>
+#include <iostream>
+
+namespace marian {
+namespace bergamot {
+
+namespace {
+
+// Generic write/read from pointers
+
+// write uses ostream, to get bytes/blob use with ostringstream(::binary)
+template <class T>
+void write(std::ostream &out, const T *data, size_t num = 1) {
+  const char *cdata = reinterpret_cast<const char *>(data);
+  out.write(cdata, num * sizeof(T));
+}
+
+// L4 stores the data as blobs. These use memcpy to construct parts from the blob given the start and size. It is the
+// responsibility of the caller to prepare the container with the correct contiguous size so memcpy works correctly.
+template <class T>
+const char *copyInAndAdvance(const char *src, T *dest, size_t num = 1) {
+  const void *vsrc = reinterpret_cast<const void *>(src);
+  void *vdest = reinterpret_cast<void *>(dest);
+  std::memcpy(vdest, vsrc, num * sizeof(T));
+  return reinterpret_cast<const char *>(src + num * sizeof(T));
+}
+
+// Specializations to read and write vectors. The format stored is [size, v_0, v_1 ... v_size]
+template <class T>
+void writeVector(std::ostream &out, std::vector<T> v) {
```

Comment:
These format conversions make me sad.  
If you want/need a flat format, then just make it a flat format.  And wrap a class around it that accesses the flat format.  Then remove the non-flat format.